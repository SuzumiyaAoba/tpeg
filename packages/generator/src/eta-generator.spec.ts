/**
 * Tests for Eta Template Engine Based Code Generator
 */

import { describe, expect, it } from "bun:test";
import {
  EtaTPEGCodeGenerator,
  generateEtaTypeScriptParser,
} from "./eta-generator";

// Import test utilities from core
import type { Expression, GrammarDefinition, RuleDefinition } from "./types";

// Simple test helper functions
function createGrammarDefinition(
  name: string,
  annotations: unknown[],
  rules: RuleDefinition[],
): GrammarDefinition {
  return {
    type: "GrammarDefinition",
    name,
    annotations,
    rules,
  };
}

function createRuleDefinition(
  name: string,
  pattern: Expression,
): RuleDefinition {
  return {
    type: "RuleDefinition",
    name,
    pattern,
  };
}

function createStringLiteral(value: string) {
  return {
    type: "StringLiteral",
    value,
  };
}

function createCharacterClass(
  ranges: Array<{ start: string; end?: string }>,
  negated: boolean,
) {
  return {
    type: "CharacterClass",
    ranges,
    negated,
  };
}

function createCharRange(start: string, end: string) {
  return { start, end };
}

function createSequence(elements: Expression[]) {
  return {
    type: "Sequence",
    elements,
  };
}

function createChoice(alternatives: Expression[]) {
  return {
    type: "Choice",
    alternatives,
  };
}

function createStar(expression: Expression) {
  return {
    type: "Star",
    expression,
  };
}

function createPlus(expression: Expression) {
  return {
    type: "Plus",
    expression,
  };
}

function createOptional(expression: Expression) {
  return {
    type: "Optional",
    expression,
  };
}

function createQuantified(expression: Expression, min: number, max?: number) {
  return {
    type: "Quantified",
    expression,
    min,
    max,
  };
}

function createIdentifier(name: string) {
  return {
    type: "Identifier",
    name,
  };
}

describe("EtaTPEGCodeGenerator", () => {
  describe("Basic Code Generation", () => {
    it("should generate simple rule with string literal", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("hello", createStringLiteral("world"))],
      );

      const generator = new EtaTPEGCodeGenerator({
        language: "typescript",
        namePrefix: "test_",
        includeTypes: true,
      });

      const result = await generator.generateGrammar(grammar);

      expect(result.code).toContain('import type { Parser } from "tpeg-core";');
      expect(result.code).toContain('import { literal } from "tpeg-core";');
      expect(result.code).toContain(
        'export const test_hello: Parser<any> = literal("world");',
      );
      expect(result.performance.templateEngine).toBe("eta");
    });

    it("should generate rule with character class", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition(
            "letter",
            createCharacterClass([createCharRange("a", "z")], false),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar, {
        namePrefix: "test_",
        includeTypes: true,
      });

      expect(result.code).toContain("charClass");
      expect(result.code).toContain("export const test_letter: Parser<any>");
      expect(result.code).toContain('{ from: "a", to: "z" }');
    });

    it("should generate multiple rules", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition("hello", createStringLiteral("hello")),
          createRuleDefinition("world", createStringLiteral("world")),
          createRuleDefinition(
            "greeting",
            createSequence([
              createIdentifier("hello"),
              createStringLiteral(" "),
              createIdentifier("world"),
            ]),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar, {
        namePrefix: "test_",
        includeTypes: true,
      });

      expect(result.code).toContain("export const test_hello: Parser<any>");
      expect(result.code).toContain("export const test_world: Parser<any>");
      expect(result.code).toContain("export const test_greeting: Parser<any>");
      expect(result.code).toContain(
        'sequence(test_hello, literal(" "), test_world)',
      );
      expect(result.exports).toEqual([
        "test_hello",
        "test_world",
        "test_greeting",
      ]);
    });
  });

  describe("Advanced Expressions", () => {
    it("should generate sequence expressions", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition(
            "sequence",
            createSequence([
              createStringLiteral("a"),
              createStringLiteral("b"),
              createStringLiteral("c"),
            ]),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.code).toContain(
        'sequence(literal("a"), literal("b"), literal("c"))',
      );
    });

    it("should generate choice expressions", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition(
            "choice",
            createChoice([
              createStringLiteral("true"),
              createStringLiteral("false"),
            ]),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.code).toContain(
        'choice(literal("true"), literal("false"))',
      );
    });

    it("should generate repetition expressions", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition("star", createStar(createStringLiteral("a"))),
          createRuleDefinition("plus", createPlus(createStringLiteral("b"))),
          createRuleDefinition(
            "optional",
            createOptional(createStringLiteral("c")),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.code).toContain('zeroOrMore(literal("a"))');
      expect(result.code).toContain('oneOrMore(literal("b"))');
      expect(result.code).toContain('optional(literal("c"))');
    });

    it("should generate quantified expressions", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [
          createRuleDefinition(
            "exactly3",
            createQuantified(createStringLiteral("a"), 3, 3),
          ),
          createRuleDefinition(
            "range2to5",
            createQuantified(createStringLiteral("b"), 2, 5),
          ),
          createRuleDefinition(
            "minimum3",
            createQuantified(createStringLiteral("c"), 3),
          ),
          createRuleDefinition(
            "zero0",
            createQuantified(createStringLiteral("d"), 0, 0),
          ),
        ],
      );

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.code).toContain('quantified(literal("a"), 3, 3)');
      expect(result.code).toContain('quantified(literal("b"), 2, 5)');
      expect(result.code).toContain('quantified(literal("c"), 3)');
      expect(result.code).toContain('quantified(literal("d"), 0, 0)');
    });
  });

  describe("Type Generation Options", () => {
    it("should generate without types when includeTypes is false", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("hello", createStringLiteral("world"))],
      );

      const result = await generateEtaTypeScriptParser(grammar, {
        includeTypes: false,
      });

      expect(result.code).toContain('export const hello = literal("world");');
      expect(result.code).not.toContain(": Parser<any>");
    });

    it("should generate with custom name prefix", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("rule", createStringLiteral("value"))],
      );

      const result = await generateEtaTypeScriptParser(grammar, {
        namePrefix: "custom_",
      });

      expect(result.code).toContain("export const custom_rule");
    });

    it("should skip imports when includeImports is false", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("hello", createStringLiteral("world"))],
      );

      const result = await generateEtaTypeScriptParser(grammar, {
        includeImports: false,
      });

      expect(result.code).not.toContain("import");
      expect(result.code).toContain("export const hello");
    });
  });

  describe("Performance Analysis", () => {
    it("should include performance metadata", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("simple", createStringLiteral("hello"))],
      );

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.performance).toBeDefined();
      expect(result.performance.templateEngine).toBe("eta");
      expect(result.performance.estimatedComplexity).toBe("low");
      expect(typeof result.performance.generationTime).toBe("number");
      expect(Array.isArray(result.performance.optimizationSuggestions)).toBe(
        true,
      );
    });
  });

  describe("Error Handling", () => {
    it("should handle empty grammar", async () => {
      const grammar = createGrammarDefinition("EmptyGrammar", [], []);

      const result = await generateEtaTypeScriptParser(grammar);

      expect(result.code).toContain("import type { Parser }");
      expect(result.exports).toEqual([]);
    });

    it("should handle invalid template directory gracefully", async () => {
      const grammar = createGrammarDefinition(
        "TestGrammar",
        [],
        [createRuleDefinition("rule", createStringLiteral("value"))],
      );

      const generator = new EtaTPEGCodeGenerator({
        language: "typescript",
        templatesDir: "/nonexistent/path",
        cache: false,
      });

      // Should still attempt to generate, even if templates are missing
      // The actual error handling depends on Eta's behavior
      await expect(generator.generateGrammar(grammar)).rejects.toThrow();
    });
  });
});
