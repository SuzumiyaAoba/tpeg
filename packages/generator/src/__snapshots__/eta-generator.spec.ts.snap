// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for simple grammar 1`] = `
"import type { Parser } from "tpeg-core";import { charClass, literal } from "tpeg-core";
export const test_hello: Parser<any> = literal("world");
export const test_number: Parser<any> = charClass([{ from: "0", to: "9" }]);
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for simple grammar 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { charClass, literal } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for simple grammar 3`] = `
[
  "test_hello",
  "test_number",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for simple grammar 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for complex grammar 1`] = `
"import type { Parser } from "tpeg-core";import { charClass, choice, literal, oneOrMore, optional, sequence } from "tpeg-core";
/**
 * Generated TPEG Parser: ComplexGrammar
 * 
 * This file was automatically generated from a TPEG grammar.
 * Do not edit this file directly - regenerate from the grammar instead.
 */
export const complex_digit: Parser<any> = charClass([{ from: "0", to: "9" }]);
export const complex_letter: Parser<any> = charClass([{ from: "a", to: "z" }]);
export const complex_word: Parser<any> = oneOrMore(charClass([{ from: "a", to: "z" }]));
export const complex_expression: Parser<any> = memoize(sequence(complex_word, optional(literal(" ")), complex_word));
// contains recursionexport const complex_number: Parser<any> = choice(oneOrMore(charClass([{ from: "0", to: "9" }])), sequence(oneOrMore(charClass([{ from: "0", to: "9" }])), literal("."), oneOrMore(charClass([{ from: "0", to: "9" }]))));
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for complex grammar 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { charClass, choice, literal, oneOrMore, optional, sequence } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for complex grammar 3`] = `
[
  "complex_digit",
  "complex_letter",
  "complex_word",
  "complex_expression",
  "complex_number",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code for complex grammar 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent optimized code 1`] = `
"import type { Parser } from "tpeg-core";import { choice, literal, sequence, zeroOrMore } from "tpeg-core";
// Performance optimizations
import { globalPerformanceMonitor } from "tpeg-generator";
/**
 * Generated TPEG Parser: OptimizedGrammar
 * 
 * This file was automatically generated from a TPEG grammar.
 * Do not edit this file directly - regenerate from the grammar instead.
 */
export const opt_recursive: Parser<any> = memoize(choice(literal("a"), sequence(literal("("), opt_recursive, literal(")"))));
// contains recursionexport const opt_highComplexity: Parser<any> = zeroOrMore(choice(literal("x"), literal("y"), literal("z")));


// Performance monitoring exports
export { globalPerformanceMonitor };"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent optimized code 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { choice, literal, sequence, zeroOrMore } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent optimized code 3`] = `
[
  "opt_recursive",
  "opt_highComplexity",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent optimized code 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without types 1`] = `
"import type { Parser } from "tpeg-core";import { choice, literal } from "tpeg-core";
export const notypes_simple = literal("value");
export const notypes_choice = choice(literal("a"), literal("b"));
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without types 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { choice, literal } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without types 3`] = `
[
  "notypes_simple",
  "notypes_choice",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without types 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without imports 1`] = `
"export const noimports_basic: Parser<any> = literal("test");
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without imports 2`] = `[]`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without imports 3`] = `
[
  "noimports_basic",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code without imports 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with custom name prefix 1`] = `
"import type { Parser } from "tpeg-core";import { literal, sequence } from "tpeg-core";
export const custom_prefix_rule1: Parser<any> = literal("value1");
export const custom_prefix_rule2: Parser<any> = literal("value2");
export const custom_prefix_rule3: Parser<any> = memoize(sequence(custom_prefix_rule1, literal(" "), custom_prefix_rule2));
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with custom name prefix 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { literal, sequence } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with custom name prefix 3`] = `
[
  "custom_prefix_rule1",
  "custom_prefix_rule2",
  "custom_prefix_rule3",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with custom name prefix 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with quantified expressions 1`] = `
"import type { Parser } from "tpeg-core";import { literal, optional, quantified } from "tpeg-core";
export const quant_exactly3: Parser<any> = quantified(literal("a"), 3, 3);
export const quant_range2to5: Parser<any> = quantified(literal("b"), 2, 5);
export const quant_minimum3: Parser<any> = quantified(literal("c"), 3);
export const quant_optional: Parser<any> = optional(literal("d"));
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with quantified expressions 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { literal, optional, quantified } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with quantified expressions 3`] = `
[
  "quant_exactly3",
  "quant_range2to5",
  "quant_minimum3",
  "quant_optional",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with quantified expressions 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with lookahead expressions 1`] = `
"import type { Parser } from "tpeg-core";import { andPredicate, literal, notPredicate } from "tpeg-core";
export const lookahead_positive: Parser<any> = andPredicate(literal("a"));
export const lookahead_negative: Parser<any> = notPredicate(literal("b"));
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with lookahead expressions 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { andPredicate, literal, notPredicate } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with lookahead expressions 3`] = `
[
  "lookahead_positive",
  "lookahead_negative",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with lookahead expressions 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with labeled expressions 1`] = `
"import type { Parser } from "tpeg-core";import { literal } from "tpeg-core";
export const labeled_labeled: Parser<any> = /* label: test_label */ literal("value");
"
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with labeled expressions 2`] = `
[
  "import type { Parser } from "tpeg-core";",
  "import { literal } from "tpeg-core";",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with labeled expressions 3`] = `
[
  "labeled_labeled",
]
`;

exports[`EtaTPEGCodeGenerator Snapshot Tests should generate consistent code with labeled expressions 4`] = `
{
  "estimatedComplexity": "low",
  "optimizationSuggestions": [],
  "templateEngine": "eta",
}
`;
